fog/view/server/src/db_fetcher.rs

```diff
@@ -495,94 +525,151 @@ mod tests {
-            assert_eq!(fetched_record.ingest_invocation_id, invoc_id1);
-            assert_eq!(fetched_record.block_index, i as u64 + 10);
+            assert_eq!(fetched_record.ingress_key, ingress_key);
+            assert_eq!(fetched_record.block_index, i as u64 + 20);
             assert_eq!(blocks_and_records[i].1, fetched_record.records);
         }
 
         assert!(db_fetcher.get_pending_fetched_records().is_empty()); // The previous call should have drained this
 
+        let (highest_known_block_index, ingress_keys, missing_block_ranges) =
+            db_fetcher.get_highest_processed_block_context();
+        assert_eq!(
+            ingress_keys,
+            vec![IngressPublicKeyRecord {
+                key: ingress_key.clone(),
+                status: IngressPublicKeyStatus {
+                    start_block: 10,
+                    pubkey_expiry: 0,
+                    retired: false,
+                },
+                last_scanned_block: Some(29),
+            }]
+        );
+        assert_eq!(highest_known_block_index, 29);
+        assert!(missing_block_ranges.is_empty());
+
         // Add more blocks but this time leave a hole between the previous blocks and the new ones.
         // They should not get picked up untill a missed blocks range is reported.
-        let mut blocks_and_records = Vec::new();
-        for i in 30..40 {
+        let mut blocks_and_records_40_50 = Vec::new();
+        for i in 40..50 {
             let (block, records) = fog_test_infra::db_tests::random_block(&mut rng, i, 5); // 5 outputs per block
 
             db.add_block_data(&invoc_id1, &block, 0, &records).unwrap();
-            blocks_and_records.push((block, records));
+            blocks_and_records_40_50.push((block, records));
         }
 
         sleep(Duration::from_secs(1)); // Supposedly enough time for at least some blocks to get picked up.
 
         assert!(db_fetcher.get_pending_fetched_records().is_empty());
 
-        // Report a missing block range.
-        db.report_missed_block_range(&BlockRange::new(20, 30))
+        // TODO The last scanned block index gets updated even though we have a hole.
+        // I am not sure this is desirable...
+        let (highest_known_block_index, ingress_keys, missing_block_ranges) =
+            db_fetcher.get_highest_processed_block_context();
+        assert_eq!(
+            ingress_keys,
+            vec![IngressPublicKeyRecord {
+                key: ingress_key.clone(),
+                status: IngressPublicKeyStatus {
+                    start_block: 10,
+                    pubkey_expiry: 0,
+                    retired: false,
+                },
+                last_scanned_block: Some(49),
+            }]
+        );
+        assert_eq!(highest_known_block_index, 49);
+        assert!(missing_block_ranges.is_empty());
+
+        // Report a missing block range. This should have no effect.
+        db.report_missed_block_range(&BlockRange::new(30, 40))
```

Comment:
The current version doesn't skip blocks that overlap with missed block ranges, instead it is only the pubkey expiry that controls stopping scanning for a given key. 

I want to check my understanding, @garbageslam is this flow correct?

1) We have an active ingress key with pubkey_expiry = 100, and it has scanned all the way to block 50.
2) Everything crashes and by the time we restarted, the network moved to block 75.
3) A new ingress key is created with start=75
4) We report missed block range 50-100, since we did advertise that we will scan until block 100 but failed
5) We set retired=true for the first ingress key, and this causes the view server to stop trying to load blocks from it

Reply:
Hm, I'm looking at https://github.com/garbageslam/fog/blob/1d924ff2a0663b432fbee854915c65ee02fb3246/fog/view/server/src/block_tracker.rs#L30-L32 and it looks like it will keep returning block `51` is the next one even if we are retired, because it is lower than the pubkey expiry. @garbageslam what are your thoughts on that?

Reply:
i think this behavior is correct -- the reason is, if the new ingress key starts at 75, Bob may have an RNG and some transactions that were scanned by that ingest say at blocks 77, 88, 99.
He may find those transactions by view key scanning, but if he later gets a TxOut in block 124, that will be the 4th output of the RNG. But from point of view of fog view protocol, he is still on the first output of his RNG. So even though those blocks are reported missing (because of a different key), fog view still needs to scan them, or else Bob will get blocked on rng value 4 and not find TxOut in block 124